chain/sub/incoming.go

```diff
@@ -283,59 +283,82 @@ func (bv *BlockValidator) validateMsgMeta(ctx context.Context, msg *types.BlockM
 	return nil
 }
 
-func (bv *BlockValidator) getMinerWorkerKey(ctx context.Context, msg *types.BlockMsg) (address.Address, error) {
-	addr := msg.Header.Miner
+func (bv *BlockValidator) checkPowerAndGetWorkerKey(ctx context.Context, bh *types.BlockHeader) (address.Address, error) {
+	addr := bh.Miner
 
 	bv.mx.Lock()
 	key, ok := bv.keycache[addr.String()]
 	bv.mx.Unlock()
-	if ok {
-		return key, nil
-	}
+	if !ok {
+		// TODO I have a feeling all this can be simplified by cleverer DI to use the API
+		ts := bv.chain.GetHeaviestTipSet()
+		st, _, err := bv.stmgr.TipSetState(ctx, ts)
+		if err != nil {
+			return address.Undef, err
+		}
+		buf := bufbstore.NewBufferedBstore(bv.chain.Blockstore())
+		cst := cbor.NewCborStore(buf)
+		state, err := state.LoadStateTree(cst, st)
+		if err != nil {
+			return address.Undef, err
+		}
+		act, err := state.GetActor(addr)
+		if err != nil {
+			return address.Undef, err
+		}
 
-	// TODO I have a feeling all this can be simplified by cleverer DI to use the API
-	ts := bv.chain.GetHeaviestTipSet()
-	st, _, err := bv.stmgr.TipSetState(ctx, ts)
-	if err != nil {
-		return address.Undef, err
-	}
-	buf := bufbstore.NewBufferedBstore(bv.chain.Blockstore())
-	cst := cbor.NewCborStore(buf)
-	state, err := state.LoadStateTree(cst, st)
-	if err != nil {
-		return address.Undef, err
-	}
-	act, err := state.GetActor(addr)
-	if err != nil {
-		return address.Undef, err
-	}
+		blk, err := bv.chain.Blockstore().Get(act.Head)
+		if err != nil {
+			return address.Undef, err
+		}
+		aso := blk.RawData()
 
-	blk, err := bv.chain.Blockstore().Get(act.Head)
-	if err != nil {
-		return address.Undef, err
+		var mst miner.State
+		err = mst.UnmarshalCBOR(bytes.NewReader(aso))
+		if err != nil {
+			return address.Undef, err
+		}
+
+		info, err := mst.GetInfo(adt.WrapStore(ctx, cst))
+		if err != nil {
+			return address.Undef, err
+		}
+
+		worker := info.Worker
+		key, err = bv.stmgr.ResolveToKeyAddress(ctx, worker, ts)
+		if err != nil {
+			return address.Undef, err
+		}
+
+		bv.mx.Lock()
+		bv.keycache[addr.String()] = key
+		bv.mx.Unlock()
 	}
-	aso := blk.RawData()
 
-	var mst miner.State
-	err = mst.UnmarshalCBOR(bytes.NewReader(aso))
+	// we check that the miner met the minimum power at the lookback tipset
+
+	baseTs, err := bv.stmgr.ChainStore().LoadTipSet(types.NewTipSetKey(bh.Parents...))
```

Comment:
The same issue raised by @magik6k applies here the same, we should be picking the heaviest tipset not the parent of whatever we received. Anyway, this was my fault for derailing the original intent of this PR, will submit something later this week.